#' Plot wrapper
#'
#' Wrapper around splice plotting function for a nexons output file.
#'
#' @param nexons_output_file
#'
#' @return
#' @export
#'
#' @examples
#' file <- system.file("extdata", "nexons_sirv5_f15.gtf", package = "nexonsAnalysis")
#' # just the splice picture
#' plot_wrapper(file, min_count = 5)
#' # including a quantitative plot
#' plot_wrapper(file, quant_plot = TRUE, min_count = 5)
plot_wrapper <- function(nexons_output_file, quant_plot = FALSE, min_count = 1){

  nexons_output <- readr::read_tsv(nexons_output_file)
  parsed_splices <- parse_nexons_gtf(nexons_output, min_count = min_count)
  draw_splice_picture(parsed_splices, quant = quant_plot)

}


#' Parse gtf output from nexons
#'
#' Takes a gtf file generated by nexons and separates the attribute column
#' into Transcript_id, Gene_id and splice_pattern
#'
#' @param nexon_gtf gtf file generated by nexons
#' @param min_count minimum count for transcript to be displayed (default: 1)
#'
#' @return tibble containing the columns "variant", "strand", "score", "Transcript_id", "Gene_id", "splice_pattern"
#' @export
#'
#' @examples
#' file <- system.file("extdata", "inst/extdata/nexons_sirv5_f15.gtf", package = "nexonsAnalysis")
#' nexons_output <- readr::read_tsv(file)
#' parsed_splices <- parse_nexons_gtf(nexons_output)
#'
#' @importFrom magrittr %>%
parse_nexons_gtf <- function(nexon_gtf, min_count = 1){
  nexon_gtf %>%
    dplyr::select(strand, score, attribute) %>%
    tidyr::separate(attribute, sep = "; ", into = c("Transcript_id", "Gene_id", "splice_pattern")) %>%
    dplyr::mutate(Transcript_id = gsub("transcript_id ", "", Transcript_id)) %>%
    dplyr::mutate(Gene_id = gsub("gene_id ", "", Gene_id)) %>%
    dplyr::mutate(splice_pattern = gsub("splicePattern ", "", splice_pattern)) %>%
    dplyr::arrange(Transcript_id) %>%
    dplyr::filter(score >= min_count) %>%
    tibble::rowid_to_column("variant")

}

#' Add exon loci
#'
#' parses splice patterns, adding 2 columns (start and end of exon) to a dataframe
#' containin the splice patterns
#'
#' @param splice_data dataframe that contains a column named splice_pattern
#'
#' @return original dataframe with additional columns (and therefore rows) for
#' start and end of each exon in the splice pattern
#'
#' @export
#'
#' @examples
#' none
#' @importFrom magrittr %>%
add_exon_loci <- function(splice_data){
  splice_data %>%
    dplyr::mutate(full_pattern=splice_pattern) %>%
    tidyr::separate_rows(splice_pattern,sep=":") %>%
    tidyr::separate(splice_pattern, into=c("start","end"), sep="-", convert=TRUE) %>%
    dplyr::mutate(end=as.integer(end)) %>%
    dplyr::mutate(end=dplyr::if_else(is.na(end),start,end))
}
